### DISCUSSION TITLE
[Run issue]: I can't correctly simulate SOBP

### DISCUSSION THREAD

**Discussion by ghost** [OpenTOPAS/OpenTOPAS#55] [Run-Problem] [answered]
### Contact Details

cecilianovawijaya01@gmail.com

### What was the issue?

Hello,

I am currently working on simulating Spread Out Bragg Peak (SOBP) using TOPAS but have encountered some difficulties. I followed the TOPAS-SOBP python code by [Velten and Tomé (Biomed Phys Eng Express, 2020)](https://github.com/cvelten/TOPAS-SOBP) but can't achieve the correct results. Despite trying different parameters and also following the topas website examples, I can't produce the SOBP correctly.

Here is an example of the parameters I obtained from the python code:
```
# ----- SOBP -----
s:Tf/BeamEnergy/Function = "Step"
dv:Tf/BeamEnergy/Times = 117
1.6 4.7 7.9 11.1 14.3 17.5 20.7 24.0 27.3 30.5 33.9 37.2 40.6 43.9 47.3 50.7 54.2 57.6 61.1 64.6 68.2 71.7 75.3 78.9 82.5 86.2 89.9 93.6 97.3 101.1 104.8 108.7 112.5 116.4 120.3 124.2 128.2 132.2 136.2 140.3 144.4 148.5 152.7 156.9 161.1 165.4 169.7 174.1 178.5 182.9 187.4 191.9 196.5 201.1 205.8 210.5 215.2 220.1 224.9 229.8 234.8 239.9 245.0 250.1 255.3 260.6 266.0 271.4 276.9 282.4 288.1 293.8 299.6 305.5 311.5 317.5 323.7 330.0 336.3 342.8 349.4 356.1 363.0 369.9 377.0 384.3 391.7 399.3 407.0 414.9 423.1 431.4 439.9 448.7 457.7 467.0 476.6 486.5 496.8 507.4 518.4 530.0 542.0 554.6 567.9 581.9 596.9 612.8 630.0 648.7 669.4 692.6 719.4 751.7 793.7 861.6 1000.0 ms
dv:Tf/BeamEnergy/Values = 117
228.1 228.3 228.5 228.7 228.9 229.0 229.2 229.4 229.6 229.8 230.0 230.2 230.4 230.6 230.8 231.0 231.2 231.4 231.6 231.8 232.0 232.2 232.3 232.5 232.7 232.9 233.1 233.3 233.5 233.7 233.9 234.1 234.3 234.5 234.7 234.9 235.0 235.2 235.4 235.6 235.8 236.0 236.2 236.4 236.6 236.8 237.0 237.1 237.3 237.5 237.7 237.9 238.1 238.3 238.5 238.7 238.8 239.0 239.2 239.4 239.6 239.8 240.0 240.2 240.4 240.5 240.7 240.9 241.1 241.3 241.5 241.7 241.9 242.0 242.2 242.4 242.6 242.8 243.0 243.2 243.4 243.5 243.7 243.9 244.1 244.3 244.5 244.7 244.8 245.0 245.2 245.4 245.6 245.8 246.0 246.1 246.3 246.5 246.7 246.9 247.1 247.2 247.4 247.6 247.8 248.0 248.2 248.4 248.5 248.7 248.9 249.1 249.3 249.5 249.6 249.8 250.0 MeV
```
This is my first beam source attempt:
```
# ----- Beam Source -----
s:So/ProtonBeam/Type				= "Beam"
s:So/ProtonBeam/Component			= "BeamPosition"
s:So/ProtonBeam/BeamParticle			= "proton"
d:So/ProtonBeam/BeamEnergy			= Tf/BeamEnergy/Value MeV
u:So/ProtonBeam/BeamEnergySpread		= 0
s:So/ProtonBeam/BeamPositionDistribution	= "Flat"
s:So/ProtonBeam/BeamPositionCutoffShape		= "Ellipse"
d:So/ProtonBeam/BeamPositionCutoffX		= 10. cm
d:So/ProtonBeam/BeamPositionCutoffY		= 10. cm
d:So/ProtonBeam/BeamPositionSpreadX		= 0.65 cm
d:So/ProtonBeam/BeamPositionSpreadY		= 0.65 cm
s:So/ProtonBeam/BeamAngularDistribution		= "None"
ic:So/ProtonBeam/NumberOfHistoriesInRun		= 10
```
I received this PDD plot:
![PDD-flipped](https://github.com/user-attachments/assets/f6a06e34-e12f-4998-8aa7-3d027d0de788)

My second attempt that somehow resulted in a Bragg Peak:
```
# ----- Beam Source -----
s:So/ProtonBeam/Type                     = "Beam"
s:So/ProtonBeam/Component                = "BeamPosition"
s:So/ProtonBeam/BeamParticle             = "proton"
dc:So/ProtonBeam/BeamEnergy              = Tf/BeamEnergy/Value MeV
u:So/ProtonBeam/BeamEnergySpread         = 0.757504
s:So/ProtonBeam/BeamPositionDistribution = "Gaussian"
s:So/ProtonBeam/BeamPositionCutoffShape  = "Ellipse"
d:So/ProtonBeam/BeamPositionCutoffX      = 10. cm
d:So/ProtonBeam/BeamPositionCutoffY      = 10. cm
d:So/ProtonBeam/BeamPositionSpreadX      = 0.65 cm
d:So/ProtonBeam/BeamPositionSpreadY      = 0.65 cm
s:So/ProtonBeam/BeamAngularDistribution  = "Gaussian"
d:So/ProtonBeam/BeamAngularCutoffX       = 90. deg
d:So/ProtonBeam/BeamAngularCutoffY       = 90. deg
d:So/ProtonBeam/BeamAngularSpreadX       = 0.0032 rad
d:So/ProtonBeam/BeamAngularSpreadY       = 0.0032 rad
ic:So/ProtonBeam/NumberOfHistoriesInRun  = 10
```
![PDD-flipped](https://github.com/user-attachments/assets/9bb813b2-db94-43ec-a2a7-fd6c1d061165)

Finally, my third attempt also resulted in a Bragg Peak:
```
# ----- Beam Source -----
s:So/ProtonBeam/Type				= "Beam"
s:So/ProtonBeam/Component			= "BeamPosition"
s:So/ProtonBeam/BeamParticle			= "proton"
dc:So/ProtonBeam/BeamEnergy			= Tf/BeamEnergy/Value MeV
u:So/ProtonBeam/BeamEnergySpread		= 0
s:So/ProtonBeam/BeamPositionDistribution	= "Flat"
s:So/ProtonBeam/BeamPositionCutoffShape		= "Rectangle"
d:So/ProtonBeam/BeamPositionCutoffX		= 2.5 cm
d:So/ProtonBeam/BeamPositionCutoffY		= 2.5 cm
d:So/ProtonBeam/BeamPositionSpreadX		= 2.5 cm
d:So/ProtonBeam/BeamPositionSpreadY		= 2.5 cm
s:So/ProtonBeam/BeamAngularDistribution		= "None"
ic:So/ProtonBeam/NumberOfHistoriesInRun		= 1000000
```
![PDD-flipped](https://github.com/user-attachments/assets/bf7e33a2-0c0a-4ea8-9caf-5a6e0c518135)

I also tried following the user guide example, but again only obtained a Bragg Peak:
```
# Set a raw SOBP
s:Tf/Energy/Function = "Step"
dv:Tf/Energy/Times = 5 20 40 60 80 100 ms
dv:Tf/Energy/Values = 5 160 155 150 145 140 MeV
s:Tf/Histories/Function = "Step"
dv:Tf/Histories/Times = 5 20 40 60 80 100 ms
iv:Tf/Histories/Values = 5 6600 2500 1750 1400 1200

s:So/Example/Type                     = "Beam"
s:So/Example/Component                = "BeamPosition"
s:So/Example/BeamParticle             = "proton"
d:So/Example/BeamEnergy               = Tf/Energy/Value MeV
u:So/Example/BeamEnergySpread         = 0.757504
s:So/Example/BeamPositionDistribution = "Flat"
s:So/Example/BeamPositionCutoffShape  = "Ellipse"
d:So/Example/BeamPositionCutoffX      = 10. mm
d:So/Example/BeamPositionCutoffY      = 10. mm
s:So/Example/BeamAngularDistribution  = "None"
i:So/Example/NumberOfHistoriesInRun   = 1000000 * Tf/Histories/Value
```
![PDD-flipped](https://github.com/user-attachments/assets/6f57aa90-89c8-47d8-acc9-4992a91b1f9a)

Could anyone assist me in identifying what I might be doing wrong? Or perhaps can anyone please let me know where can I find a guideline or tutorial for SOBP?

Thank you for your attention

### TOPAS category

Physics

### Version

3.9

### What Operative System (OS) are you running on?

Linux

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Attestation

- [X] I testify that I have fully read the bug reporting guideline beforehand.

**Comment by JoseRamosMendez** [comment, marked as answer]
Are you also adjusting the sequential times interval and number (for your last trial)?

    i:Tf/NumberOfSequentialTimes = 5
    d:Tf/TimelineEnd = 100

**Reply by ghost** [reply]
Ah, I didn't include that. Thank you for your response.

Here is the PDD result that I received after putting those 2 parameters to my last trial. Is this correct?
![PDD-flipped](https://github.com/user-attachments/assets/9f94b3fc-9973-48e4-a3a3-f6f0ed08d13c)

Additionally, if you don't mind, could you please briefly explain what these two parameters do? I have read the explanation on the website, but I am still unsure. Do I need to always include them when creating SOBP?

**Comment by tracyfhy** [comment]
i:Tf/NumberOfSequentialTimes = 5

number of simulations. In your case, number of bragg peaks you want to simulate.

d:Tf/TimelineEnd = 100

the time when simulation is ended.
It is related to this line.
dv:Tf/Energy/Times = 5 20 40 60 80 100 ms

But you can certainly try to change the those two to select the time interval you want to simulate.
